pkg/build/controller/build/build_controller.go

```diff
@@ -0,0 +1,824 @@
+		}
+
+		var tag string
+		streamName := outputTo.Name
+		if outputTo.Kind == "ImageStreamTag" {
+			var ok bool
+			streamName, tag, ok = imageapi.SplitImageStreamTag(streamName)
+			if !ok {
+				return "", fmt.Errorf("the referenced image stream tag is invalid: %s", outputTo.Name)
+			}
+			tag = ":" + tag
+		}
+		stream, err := bc.imageStreamStore.ImageStreams(namespace).Get(streamName, metav1.GetOptions{})
+		if err != nil {
+			if errors.IsNotFound(err) {
+				return "", fmt.Errorf("the referenced output image stream %s/%s does not exist", namespace, streamName)
+			}
+			return "", fmt.Errorf("the referenced output image stream %s/%s could not be found by build %s/%s: %v", namespace, streamName, build.Namespace, build.Name, err)
+		}
+		if len(stream.Status.DockerImageRepository) == 0 {
+			e := fmt.Errorf("the image stream %s/%s cannot be used as the output for build %s/%s because the integrated Docker registry is not configured and no external registry was defined", namespace, outputTo.Name, build.Namespace, build.Name)
+			bc.recorder.Eventf(build, kapi.EventTypeWarning, "invalidOutput", "Error starting build: %v", e)
+			return "", e
+		}
+		ref = fmt.Sprintf("%s%s", stream.Status.DockerImageRepository, tag)
+	}
+	return ref, nil
+}
+
+// createBuildPod creates a new pod to run a build
+func (bc *BuildController) createBuildPod(build *buildapi.Build) (*buildUpdate, error) {
+
+	update := &buildUpdate{}
+
+	// Set the output Docker image reference.
+	ref, err := bc.resolveOutputDockerImageReference(build)
+	if err != nil {
+		// If we cannot resolve the output reference, the output image stream
+		// may not yet exist. The build should remain in the new state and show the
+		// reason that it is still in the new state.
+		update.setReason(buildapi.StatusReasonInvalidOutputReference)
+		update.setMessage(buildapi.StatusMessageInvalidOutputRef)
+		return update, err
+	}
+
+	// Create the build pod spec
+	buildPod, err := bc.createPodSpec(build, ref)
+	if err != nil {
+		// If an error occurred when creating the pod spec, it likely means
+		// that the build is something we don't understand. For example, it could
+		// have a strategy that we don't recognize. It will remain in New state
+		// and be updated with the reason that it is still in New
+		update.setReason(buildapi.StatusReasonCannotCreateBuildPodSpec)
+		update.setMessage(buildapi.StatusMessageCannotCreateBuildPodSpec)
+
+		// The error will be logged, but will not be returned to the caller
+		// to be retried. The reason is that there's really no external factor
+		// that could cause the pod creation to fail; therefore no reason
+		// to immediately retry processing the build.
+		//
+		// A scenario where this would happen is that we've introduced a
+		// new build strategy in the master, but the old version of the controller
+		// is still running. We don't want the old controller to move the
+		// build to the error phase and we don't want it to keep actively retrying.
```

Comment:
Yes it will cause another loop through here when we update it, and we will see the build again on a relist. So not ideal.